File: chapter9/AngularShop/AngularShop/Models/OrderItem.cs
using Newtonsoft.Json;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AngularShop.Models
{
    public class OrderItem
    {
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public long Id { get; set; }
        [Required]
        public int Qty { get; set; }
        [Required]
        public decimal Price { get; set; }

        [ForeignKey("Product")]
        public long ProductId { get; set; }
        [ForeignKey("Order")]
        public long OrderId { get; set; }

        #region 프론트엔드에만 전달할 속성

        [NotMapped]
        public string ProductName
        {
            get
            {
                return (Product != null)
                    ? Product.Name
                    : _productName;
            }
            set
            {
                _productName = value;
            }
        }
        private string _productName;

        #endregion

        [JsonIgnore]
        public virtual Product Product { get; set; }
        [JsonIgnore]
        public virtual Order Order { get; set; }
    }
}

File: chapter9/AngularShop/AngularShopBoot/src/main/kotlin/com/angularShop/angularShop/models/OrderItem.kt
package com.angularShop.angularShop.models

import com.fasterxml.jackson.annotation.JsonIdentityInfo
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.ObjectIdGenerators
import java.math.BigDecimal
import javax.persistence.*

@Entity
@Table(name = "OrderItems")
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator::class, property = "id")
data class OrderItem(
        @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Long? = null,

        var qty: Int,

        var price: BigDecimal,

        @Column(insertable = false, updatable = false)
        var productId: Long? = null,

        @Column(insertable = false, updatable = false)
        var orderId: Long? = null,

        @JsonIgnore
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "productId")
        var product: Product? = null,

        @JsonIgnore
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "orderId")
        var order: Order? = null
) {
    fun getProductName() : String? {
        return product?.name ?: null
    }
}